sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala

```diff
@@ -0,0 +1,353 @@
+            cluster.id,
+            cluster.name,
+            cluster.desc,
+            toStoredNodes(cluster.nodes),
+            cluster.metrics)
+          new SparkPlanGraphNodeWrapper(null, storedCluster)
+
+        case node =>
+          new SparkPlanGraphNodeWrapper(node, null)
+      }
+    }
+
+    val planGraph = SparkPlanGraph(sparkPlanInfo)
+    val sqlPlanMetrics = planGraph.allNodes.flatMap { node =>
+      node.metrics.map { metric => (metric.accumulatorId, metric) }
+    }.toMap.values.toList
+
+    val graphToStore = new SparkPlanGraphWrapper(
+      executionId,
+      toStoredNodes(planGraph.nodes),
+      planGraph.edges)
+    kvstore.write(graphToStore)
+
+    val exec = getOrCreateExecution(executionId)
+    exec.description = description
+    exec.details = details
+    exec.physicalPlanDescription = physicalPlanDescription
+    exec.metrics = sqlPlanMetrics
+    exec.submissionTime = time
+    update(exec)
+  }
+
+  private def onExecutionEnd(event: SparkListenerSQLExecutionEnd): Unit = {
+    val SparkListenerSQLExecutionEnd(executionId, time) = event
+    liveExecutions.get(executionId).foreach { exec =>
+      synchronized {
+        exec.metricsValues = aggregateMetrics(exec)
+
+        // Remove stale LiveStageMetrics objects for stages that are not active anymore.
+        val activeStages = liveExecutions.values.flatMap { other =>
+          if (other != exec) other.stages else Nil
+        }.toSet
+        stageMetrics.retain { case (id, _) => activeStages.contains(id) }
+
+        exec.completionTime = Some(new Date(time))
+        exec.endEvents += 1
+
+        update(exec)
+      }
+    }
+  }
+
+  private def onDriverAccumUpdates(event: SparkListenerDriverAccumUpdates): Unit = {
+    val SparkListenerDriverAccumUpdates(executionId, accumUpdates) = event
+    liveExecutions.get(executionId).foreach { exec =>
+      exec.driverAccumUpdates = accumUpdates.toMap
+      update(exec)
+    }
+  }
+
+  override def onOtherEvent(event: SparkListenerEvent): Unit = event match {
+    case e: SparkListenerSQLExecutionStart => onExecutionStart(e)
+    case e: SparkListenerSQLExecutionEnd => onExecutionEnd(e)
+    case e: SparkListenerDriverAccumUpdates => onDriverAccumUpdates(e)
+    case _ => // Ignore
+  }
+
+  private def getOrCreateExecution(executionId: Long): LiveExecutionData = {
+    liveExecutions.getOrElseUpdate(executionId, new LiveExecutionData(executionId))
+  }
+
+  private def update(exec: LiveExecutionData): Unit = {
+    val now = System.nanoTime()
+    if (exec.endEvents >= exec.jobs.size + 1) {
+      liveExecutions.remove(exec.executionId)
+      exec.write(kvstore, now)
+    } else if (liveUpdatePeriodNs >= 0) {
+      if (now - exec.lastWriteTime > liveUpdatePeriodNs) {
+        exec.write(kvstore, now)
+      }
+    }
+  }
+
+  private def isSQLStage(stageId: Int): Boolean = {
+    liveExecutions.values.exists { exec =>
+      exec.stages.contains(stageId)
```

Comment:
any reason not to make `exec.stages` a `Set`?  I guess it would normally be small so it doesn't matter, but seems like it couldn't hurt to protect against some strange SQL query with some huge number of stages.